src/rhsmlib/services/register.py

```diff
@@ -0,0 +1,100 @@
+
+        options = {
+            'activation_keys': activation_keys,
+            'environment': environment,
+            'force': force,
+            'name': name,
+            'consumerid': consumerid
+        }
+        self.validate_options(options)
+
+        environment = options['environment']
+        facts_dict = self.facts.get_facts()
+
+        # Default to the hostname if no name is given
+        consumer_name = options['name'] or socket.gethostname()
+
+        self.plugin_manager.run("pre_register_consumer", name=consumer_name, facts=facts_dict)
+
+        if consumerid:
+            consumer = self.cp.getConsumer(consumerid)
+            if consumer.get('type', {}).get('manifest', {}):
+                raise exceptions.ServiceError(
+                    "Registration attempted with a consumer ID that is not of type 'system'")
+        else:
+            consumer = self.cp.registerConsumer(
+                name=consumer_name,
+                facts=facts_dict,
+                owner=org,
+                environment=environment,
+                keys=options.get('activation_keys'),
+                installed_products=self.installed_mgr.format_for_server(),
+                content_tags=self.installed_mgr.tags
+            )
+        self.installed_mgr.write_cache()
+        self.plugin_manager.run("post_register_consumer", consumer=consumer, facts=facts_dict)
+        return consumer
+
+    def validate_options(self, options):
+        error_msg = None
+
+        if self.identity.is_valid() and options['force'] is not True:
+            error_msg = _("This system is already registered. Add force to options to override.")
+        elif options.get('name') == '':
+            error_msg = _("Error: system name can not be empty.")
+        elif options['consumerid'] and options['force'] is True:
+            error_msg = _("Error: Can not force registration while attempting to recover registration with consumerid. Please use --force without --consumerid to re-register or use the clean command and try again without --force.")
+
+        # If 'activation_keys' already exists in the dictionary, leave it.  Otherwise, set to None.
+        if options['activation_keys']:
+            # 746259: Don't allow the user to pass in an empty string as an activation key
+            if '' == options['activation_keys']:
+                error_msg = _("Error: Must specify an activation key")
+            elif getattr(self.cp, 'username', None) or getattr(self.cp, 'password', None):
+                error_msg = _("Error: Activation keys do not require user credentials.")
+            elif options['consumerid']:
+                error_msg = _("Error: Activation keys can not be used with previously registered IDs.")
+            elif options['environment']:
+                error_msg = _("Error: Activation keys do not allow environments to be specified.")
+        elif not getattr(self.cp, 'username', None) or not getattr(self.cp, 'password', None):
```

Comment:
In the case that you're already registered, this branch overrides the more accurate error message already set ("This system is already registered. Add force to options to override."), as `build_uep` gives a connection using the identity cert if it exists, regardless of whether username/password is passed.